ffi/tests/firewood/merkle_compatibility_test.go

```diff
@@ -0,0 +1,442 @@
+package tests
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"path/filepath"
+	"slices"
+	"testing"
+
+	firewood "github.com/ava-labs/firewood-go/ffi"
+
+	"github.com/ava-labs/avalanchego/database"
+	"github.com/ava-labs/avalanchego/database/memdb"
+	"github.com/ava-labs/avalanchego/x/merkledb"
+	"github.com/stretchr/testify/require"
+)
+
+const (
+	keySize      = 4
+	valSize      = 4
+	maxBatchSize = 5
+	maxNumKeys   = 10_000
+)
+
+const (
+	dbGet byte = iota
+	dbUpdate
+	dbBatch
+	checkDBHash
+	createProposalOnProposal
+	createProposalOnDB
+	proposalGet
+	commitProposal
+	maxStep
+)
+
+var stepMap = map[byte]string{
+	dbGet:                    "dbGet",
+	dbUpdate:                 "dbUpdate",
+	dbBatch:                  "dbBatch",
+	checkDBHash:              "checkDBHash",
+	createProposalOnProposal: "createProposalOnProposal",
+	createProposalOnDB:       "createProposalOnDB",
+	proposalGet:              "proposalGet",
+	commitProposal:           "commitProposal",
+}
+
+func newTestDatabase(t *testing.T) *firewood.Database {
+	t.Helper()
+
+	dbFile := filepath.Join(t.TempDir(), "test.db")
+	db, closeDB, err := newDatabase(dbFile)
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		require.NoError(t, closeDB())
+	})
+	return db
+}
+
+func newDatabase(dbFile string) (*firewood.Database, func() error, error) {
+	conf := firewood.DefaultConfig()
+	conf.MetricsPort = 0
+	conf.Create = true
+
+	f, err := firewood.New(dbFile, conf)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create new database at filepath %q: %w", dbFile, err)
+	}
+	return f, f.Close, nil
+}
+
+type tree struct {
+	require *require.Assertions
+	rand    *rand.Rand
+
+	id       int
+	nextID   int
+	merkleDB merkledb.MerkleDB
+	fwdDB    *firewood.Database
+
+	children []*proposal
+
+	proposals map[int]*proposal
+	keys      [][]byte
+}
+
+type proposal struct {
+	parentID   int
+	id         int
+	merkleView merkledb.View
+	fwdView    *firewood.Proposal
+
+	children []*proposal
+}
+
+func newTestTree(t *testing.T, rand *rand.Rand) *tree {
+	r := require.New(t)
+
+	memdb := memdb.New()
+	merkleDB, err := merkledb.New(context.Background(), memdb, merkledb.NewConfig())
+	r.NoError(err)
+
+	tr := &tree{
+		merkleDB:  merkleDB,
+		fwdDB:     newTestDatabase(t),
+		proposals: make(map[int]*proposal),
+		children:  make([]*proposal, 0),
+		require:   r,
+		rand:      rand,
+		keys:      createRandomByteSlices(min(1, rand.Intn(maxNumKeys)), keySize, rand),
+	}
+	return tr
+}
+
+func (tr *tree) selectRandomProposal() *proposal {
+	if len(tr.proposals) == 0 {
+		return nil
+	}
+
+	proposalIDs := make([]int, 0, len(tr.proposals))
+	for proposalID := range tr.proposals {
+		proposalIDs = append(proposalIDs, proposalID)
+	}
+	slices.Sort(proposalIDs)
+	selectedProposalID := tr.rand.Intn(len(proposalIDs))
+	return tr.proposals[selectedProposalID]
+}
+
+func (tr *tree) dropAllProposals() {
+	// Drop all firewood proposals out of order (allowed by firewood).
```

Comment:
... but not by merkleDB? Curious as to why you mentioned this was allowed.

Reply:
MerkleDB does not require explicitly dropping merkle views as they are simply garbage collected, but firewood requires they are explicitly dropped, so that the FFI can tell Firewood when to drop the associated memory.

It's non-obvious to me that it would be safe to drop them out of order given they are stacked on top of each other, so I wanted to put a comment since I drop them in a random order by iterating over the proposal map here (maps provide random iteration in go). Since there's no need to drop merkle proposals at all, I didn't mention MerkleDB here.